bpfd: Add new service that pulls and logs bpf program/map information

arthurshau opened the pull request:
BPF service, a.k.a. bpfd, is a new service that’s meant to serve the needs of deploying, monitoring, auditing, and facilitating the  adoption of BPF kernel technology.

Some of the general functions that bpfd might aim to provide:
- Monitoring BPF applications (programs, maps, links, etc), their resource utilization), information such as owner process/user, attach parameters (e.g., TC priority) and relationship between each other and deployed services.
- Auditing bpf() syscall usage by various users to prevent misuse and abuse
- Simplifying and improving BPF logistics and adoption

For now, the service code simply gathers data about all currently attached BPF programs and maps and logs it to journald (one entry per program/map). This includes info such as program/map name, program/map type, memlock, etc.

Things I'd like feedback about:
- Maybe aggregating all the entries into one single log entry would be better? So that users have an easier time consuming it.
- How best to proceed with this current PR - if you think the functionality is enough to merge as is. 
- Thoughts about BPF service in general and any features you would like to see added
- Comments about style/efficiency, as I'm relatively newer to C and systemd

cc: @DaanDeMeyer @anitazha

























































































<!-- devel-freezer = {"comment-id":"1633306755","freezing-tag":"v256-rc3"} -->

lperkov commented:
Thank you @arthurshau, much appreciated! Hopefully this will be a good foundation for the discussion we had here https://github.com/systemd/systemd/issues/26510

jordalgo commented on src/bpf/bpfd-manager.c:
```diff
@@ -0,0 +1,342 @@
+
+static int get_memlock_by_fd(int fd, char **memlock) {
+        _cleanup_free_ char *fdinfo = NULL;
+        char *p;
```
naming nit: Maybe "memlock_line" ?

jordalgo commented on src/bpf/bpfd-manager.c:
```diff
@@ -0,0 +1,342 @@
+                /* Grab the FD using the ID*/
+                fd = sym_bpf_map_get_fd_by_id(id);
+                if (fd < 0)
+                        return log_error_errno(errno, "Can't get fd for id: %u, ignoring", id);
```
If we're just ignoring here do we want to return or rather just log and continue iterating?

arthurshau replied:
I think we can just return, I put "ignoring" while testing but forgot to remove it.

arthurshau replied:
After more discussion, we've decided that just continuing will probably be OK. Will update most of the error handling in the code to reflect this.

jordalgo commented on src/bpf/bpfd-manager.c:
```diff
@@ -0,0 +1,342 @@
+                /* Fetch bpf_map_info using fd */
+                r = sym_bpf_obj_get_info_by_fd(fd, &map_info, &map_info_len);
+                if (r < 0)
+                        return log_error_errno(errno, "Can't get info for fd: %i", fd);
```
Same question as above.

jordalgo commented on src/bpf/bpfd-manager.h:
```diff
@@ -0,0 +1,24 @@
+#include "macro.h"
+
+/* Polling interval for BPF timer  */
+#define BPF_TIMER_INTERVAL_USEC 5 * USEC_PER_SEC /* 5.0 seconds */
```
Should we make this configurable?

arthurshau replied:
I've seen other services define their timer intervals in *-manager.h like this, so I think this is consistent with the style. I will increase the interval though since I forgot - 5 seconds is too short.

arthurshau commented:
> Would it be better to expose these items via D-Bus rather than log them? As a result, this information could be consumed in a more flexible manner.

We've considered the possibility of using D-Bus or Varlink to expose an API so that users can grab this info on-demand rather than logging to journald on an interval like it is now. I don't have a strong preference either way so I'll defer to what the maintainers would like.

arthurshau commented:
Added varlink API, accessible by using bpfctl. Also planning to add a config option to disable logging to journal if the user doesn't want that.

poettering commented on src/bpf/bpfctl.c:
```diff
@@ -0,0 +1,220 @@
+                static const JsonDispatch dispatch_table[] = {
+                        { "NAME",  JSON_VARIANT_STRING, json_dispatch_string, offsetof(struct bpf_map_data, name),  JSON_MANDATORY },
+                        { "TYPE", JSON_VARIANT_UNSIGNED, json_dispatch_uint64, offsetof(struct bpf_map_data, type), JSON_MANDATORY},
+                        { "MEMLOCK", JSON_VARIANT_STRING, json_dispatch_string, offsetof(struct bpf_map_data, memlock), JSON_MANDATORY},
```
hmm, uppercase fields in json? that's … unusual …

poettering commented on src/bpf/bpfctl.c:
```diff
@@ -0,0 +1,220 @@
+
+                r = json_dispatch(v, dispatch_table, NULL, 0, &p);
+                if (r < 0)
+                        return r;
```
this would leak the strings allocated on the heap that are stored in `bpf_map_data`.

please add a destructor call `bpf_map_data_done()` that frees the fields of the structure. Then use `_cleanup_(bpf_map_data_done() struct bpf_map_data p = {};` to allocate the struct and free it automatically when it goes out of scope.

poettering commented on src/bpf/bpfctl.c:
```diff
@@ -0,0 +1,220 @@
+                        {}
+                };
+                struct bpf_prog_data p = {
+                        .map_names = NULL,
```
this line is redundant, zero initialization is defualt, you can just do `= {}` here

poettering commented on src/bpf/bpfctl.c:
```diff
@@ -0,0 +1,220 @@
+                };
+                struct bpf_prog_data p = {
+                        .map_names = NULL,
+                };
```
leaks memory, as above

poettering commented on src/bpf/bpfctl.c:
```diff
@@ -0,0 +1,220 @@
+        r = varlink_bind_reply(vl, query_bpf_progs_and_maps_reply);
+        if (r < 0)
+                return log_debug_errno(r, "Failed to bind reply callback: %m");
+        char method[] = "io.systemd.BpfProgsAndMaps.GetBpfProgsAndMaps";
```
why the indirection? you can just pass a literal string to varlink_observe()

poettering commented on src/bpf/bpfctl.c:
```diff
@@ -0,0 +1,220 @@
+
+        r = sd_event_loop(event);
+        if (r < 0)
+                return log_error_errno(r, "Failed to run event loop: %m");
```
if there's no strong reason to use an event loop here (i.e. that you have multiple events to care for) then I'd just lop around varlink_process() here

arthurshau replied:
Agreed - ideally using varlink_collect here will do the trick.

poettering commented on src/bpf/bpfctl.c:
```diff
@@ -0,0 +1,220 @@
+
+        int r;
+        JsonVariant *m, *v;
+
```
btw, the `flags` param will tell you when you are done, i.e. if the VARLINK_REPLY_CONTINUE bit is unset this was the last reply

arthurshau replied:
Same as above

poettering commented on src/bpf/bpfd-manager.c:
```diff
@@ -0,0 +1,448 @@
+#include "bpf-dlopen.h"
+#include "fileio.h"
+#include "sd-daemon.h"
+#include "sd-messages.h"
```
please put `sd-*.h` headers first.

(they are out public API, and hence sit somewhere between the public includes we do via `#include <…>` and our local includes we do via `#include "…"`)

poettering commented on src/bpf/bpfd-manager.c:
```diff
@@ -0,0 +1,448 @@
+        memlock_line += strspn(memlock_line, WHITESPACE);
+        memlock_line[strcspn(memlock_line, WHITESPACE)] = 0;
+
+        *memlock = strdup(memlock_line);
```
coding style:

we have this rule that return params should not be touched on failure, and always be initialized on success. here you assign a value to the return parameter even on failure.

hence please, store result of strdup() in temporary var first, and only if it succeeded copy to the return parameter

poettering commented on src/bpf/bpfd-manager.c:
```diff
@@ -0,0 +1,448 @@
+#include "stdio-util.h"
+#include "umask-util.h"
+
+static int get_memlock_by_fd(int fd, char **memlock) {
```
we have this coding style rule that return parameters should be called `ret` or `ret_xyz`. the 2nd parameter here is a return param, hence please rename to `ret` or to `ret_memlock` please

poettering commented on src/bpf/bpfd-manager.c:
```diff
@@ -0,0 +1,448 @@
+
+                if (link) {
+                        r = json_build(&v, JSON_BUILD_OBJECT(
+                                                JSON_BUILD_PAIR_STRING("TYPE", "MAP")));
```
oom check missing

poettering commented on src/bpf/bpfd-manager.c:
```diff
@@ -0,0 +1,448 @@
+                                                JSON_BUILD_PAIR_OBJECT("DATA",
+                                                JSON_BUILD_PAIR_STRING("NAME", map_info.name),
+                                                JSON_BUILD_PAIR_UNSIGNED("TYPE", map_info.type),
+                                                JSON_BUILD_PAIR_STRING("MEMLOCK", memlock))));
```
why build this object in two steps?

(oom check missing)

poettering commented on src/bpf/bpfd-manager.c:
```diff
@@ -0,0 +1,448 @@
+                                                JSON_BUILD_PAIR_UNSIGNED("TYPE", map_info.type),
+                                                JSON_BUILD_PAIR_STRING("MEMLOCK", memlock))));
+
+                        varlink_notify(link, v);
```
error handling...

poettering commented on src/bpf/bpfd-manager.c:
```diff
@@ -0,0 +1,448 @@
+                                "NAME=%s", map_info.name,
+                                "TYPE=%u", map_info.type,
+                                "MEMLOCK=%s", memlock);
+                }
```
coding style: no {} around single line if/else blocks

poettering commented on src/bpf/bpfd-manager.c:
```diff
@@ -0,0 +1,448 @@
+                /* Fetch bpf_prog_info using fd */
+                r = sym_bpf_obj_get_info_by_fd(fd, &prog_info, &prog_info_len);
+                if (r < 0) {
+                        log_error("Can't get bpf prog info for fd: %i", fd);
```
this returns a useful error code, doesn't it? please use log_error_errno() in that case, and include error string via `%m` in output

poettering commented on src/bpf/bpfd-manager.c:
```diff
@@ -0,0 +1,448 @@
+                 * to get a list of map ids */
+                nr_map_ids = prog_info.nr_map_ids;
+                if (nr_map_ids > 0) {
+                        prog_map_ids = malloc(nr_map_ids * sizeof(uint32_t));
```
use type-safe and integer-overrun-safe `new(uint32_t, nr_map_ids)` please

poettering commented on src/bpf/bpfd-manager.c:
```diff
@@ -0,0 +1,448 @@
+                        prog_map_ids = malloc(nr_map_ids * sizeof(uint32_t));
+                        if (!prog_map_ids)
+                                return -ENOMEM;
+                        prog_map_names = malloc(nr_map_ids * BPF_OBJ_NAME_LEN);
```
ditto, and everywhere else

poettering commented on src/bpf/bpfd-manager.c:
```diff
@@ -0,0 +1,448 @@
+
+                        zero(prog_info);
+                        prog_info.nr_map_ids = nr_map_ids;
+                        prog_info.map_ids = PTR_TO_UINT64(prog_map_ids);
```
we usually prefer structured initialization (except for very few cases where this doesn't work because union semantics are weird). ii.e.

```c
prog_info = (struct bpf_prog_info) {
        .nr_map_ids = nr_map_ids,
        .map_ids = PTR_TO_UINT64(prog_map_ids),
};
```

poettering commented on src/bpf/bpfd-manager.c:
```diff
@@ -0,0 +1,448 @@
+                _cleanup_free_ char *memlock = NULL;
+                _cleanup_free_ uint32_t *prog_map_ids = NULL;
+                _cleanup_free_ char *prog_name = NULL;
+                _cleanup_free_ char *prog_map_names = NULL;
```
please merge lines with the same types

arthurshau commented:
Added usage of varlink_collect and added configuration to toggle logging of BPF progs and maps to journal.

DaanDeMeyer commented on src/bpf/bpfctl.c:
```diff
@@ -0,0 +1,215 @@
+typedef struct BpfProgData BpfProgData;
+typedef struct BpfMapData BpfMapData;
+
+struct BpfProgData {
```
We tend to not abbreviate too much, so please use `BpfProgramData`  here.

DaanDeMeyer commented on src/bpf/bpfctl.c:
```diff
@@ -0,0 +1,215 @@
+                        if (r < 0)
+                                return r;
+
+                        log_info(" prog_name: %s, prog_type: %lu, memlock: %s, prog_tag: %s, prog_map_names: %s, run_time_ns: %lu, run_cnt: %lu", p.name, p.type, p.memlock, p.tag, p.map_names, p.run_time_ns, p.run_cnt);
```
This should be formatted in a table using the helpers from format-table.h

DaanDeMeyer commented on src/bpf/bpfctl.c:
```diff
@@ -0,0 +1,215 @@
+        JsonVariant *i = NULL, *m = NULL, *v = NULL;
+        int r;
+
+        r = sd_event_default(&event);
+        if (r < 0)
+                return log_error_errno(r, "Failed to get event loop: %m");
```
The event loop isn't being used for anything?

DaanDeMeyer commented on src/bpf/bpfctl.c:
```diff
@@ -0,0 +1,215 @@
+
+        r = varlink_connect_address(&vl, "/run/systemd/bpf/io.systemd.Bpf");
+        if (r < 0)
+                return log_debug_errno(r, "Unable to connect to /run/systemd/bpf/io.systemd.Bpf: %m");
```
In applications we don't need to do debug logging so you can use `log_error_errno()` here instead.

DaanDeMeyer commented on src/bpf/bpfd-manager.c:
```diff
@@ -0,0 +1,489 @@
+
+        r = varlink_server_bind_method(
+                        m->varlink_server,
+                        "io.systemd.BpfProgsAndMaps.GetBpfProgsAndMaps", vl_method_get_bpf_progs_and_maps
```
We should probably start with a generic `Describe` method so we can add more stuff here later if we want. If needed we can add more specific methods as well that only return some specific type of data later.

DaanDeMeyer commented on src/bpf/bpfd-manager.c:
```diff
@@ -0,0 +1,489 @@
+        (void) mkdir_p("/run/systemd/bpf", 0755);
+        socket_path = "/run/systemd/bpf/io.systemd.Bpf";
+
+        r = varlink_server_listen_address(m->varlink_server, socket_path, 0666);
```
I wonder if we should make this accessible to everyone. Don't know enough about BPF to say whether this could be security sensitive data or not.

DaanDeMeyer commented on src/bpf/bpfd-manager.c:
```diff
@@ -0,0 +1,489 @@
+}
+
+static int log_bpf_maps_and_progs(Manager *m) {
+        int r = fetch_bpf_maps_and_progs(NULL);
+        if (r < 0)
+                return r;
+        return 0;
```
This can be simplified to `return fetch_bpf_maps_and_progs(NULL)`

DaanDeMeyer commented on src/bpf/bpfd-manager.c:
```diff
@@ -0,0 +1,489 @@
+        }
+
+        char *map_name;
+        HASHMAP_FOREACH(map_name, map_name_by_id) {
+                free(map_name);
+        }
```
No braces along single line if statements.

DaanDeMeyer commented on src/bpf/bpfd-manager.c:
```diff
@@ -0,0 +1,489 @@
+                                continue;
+                        }
+                } else {
+                        log_struct(LOG_INFO,
+                                LOG_MESSAGE("Logging info about BPF map: %s", map_info.name),
+                                "MESSAGE_ID=" SD_MESSAGE_BPFD_LOG_STR,
+                                "NAME=%s", map_info.name,
+                                "TYPE=%u", map_info.type,
+                                "MEMLOCK=%s", memlock);
```
Instead of logging a single line per BPF program/map we should log the JSON object containing the data on all maps and programs once. We'd construct the same JSON object whether we're doing varlink or logging to the journal, the only difference should be where we write the JSON object to.

DaanDeMeyer commented on src/bpf/bpfd-manager.c:
```diff
@@ -0,0 +1,489 @@
+                /* Grab the FD using the ID*/
+                fd = sym_bpf_map_get_fd_by_id(id);
+                if (fd < 0) {
+                        log_error("Can't get fd for map id: %u", id);
```
Always include the specific error when some operation fails (I'm not 100% sure how libbpf communicates its errors). Also, if you ignore an error, it should be something like `log_error_errno(fd, "Failed to get fd for map id, ignoring: %m");`

DaanDeMeyer commented on src/bpf/bpfd-manager.c:
```diff
@@ -0,0 +1,489 @@
+        xsprintf(path, "/proc/self/fdinfo/%i", fd);
+        r = read_full_virtual_file(path, &fdinfo, NULL);
+        if (r < 0)
+                return r;
```
None of this code is library code so please complain loudly about every error to make this easier to debug.

DaanDeMeyer commented on src/bpf/bpfd-manager.c:
```diff
@@ -0,0 +1,489 @@
+
+                /* Format prog tag */
+                prog_tag[BPF_TAG_SIZE * 2] = '\0';
+                for (size_t j = 0; j < BPF_TAG_SIZE; ++j) {
```
No braces for single line for loops

yuwata commented:
```
../src/bpf/bpfd-manager.c:53:29: error: variable has incomplete type 'struct bpf_map_info'
        struct bpf_map_info map_info = {};
                            ^
../src/bpf/bpfd-manager.c:53:16: note: forward declaration of 'struct bpf_map_info'
        struct bpf_map_info map_info = {};
               ^
../src/bpf/bpfd-manager.c:55:30: error: variable has incomplete type 'struct bpf_prog_info'
        struct bpf_prog_info prog_info = {};
                             ^
../src/bpf/bpfd-manager.c:55:16: note: forward declaration of 'struct bpf_prog_info'
        struct bpf_prog_info prog_info = {};
               ^
../src/bpf/bpfd-manager.c:58:30: error: variable has incomplete type 'struct bpf_func_info'
        struct bpf_func_info func_info = {};
                             ^
../src/bpf/bpfd-manager.c:58:16: note: forward declaration of 'struct bpf_func_info'
        struct bpf_func_info func_info = {};
               ^
../src/bpf/bpfd-manager.c:62:23: error: use of undeclared identifier 'BPF_TAG_SIZE'
        char prog_tag[BPF_TAG_SIZE * 2 + 1];
                      ^
../src/bpf/bpfd-manager.c:70:16: error: implicit declaration of function 'sym_bpf_map_get_next_id' is invalid in C99 [-Werror,-Wimplicit-function-declaration]
        while (sym_bpf_map_get_next_id(id, &id) >= 0) {
               ^
../src/bpf/bpfd-manager.c:79:22: error: implicit declaration of function 'sym_bpf_map_get_fd_by_id' is invalid in C99 [-Werror,-Wimplicit-function-declaration]
                fd = sym_bpf_map_get_fd_by_id(id);
                     ^
../src/bpf/bpfd-manager.c:79:22: note: did you mean 'sym_bpf_map_get_next_id'?
../src/bpf/bpfd-manager.c:70:16: note: 'sym_bpf_map_get_next_id' declared here
        while (sym_bpf_map_get_next_id(id, &id) >= 0) {
               ^
../src/bpf/bpfd-manager.c:86:21: error: implicit declaration of function 'sym_bpf_obj_get_info_by_fd' is invalid in C99 [-Werror,-Wimplicit-function-declaration]
                r = sym_bpf_obj_get_info_by_fd(fd, &map_info, &map_info_len);
                    ^
../src/bpf/bpfd-manager.c:86:21: note: did you mean 'sym_bpf_map_get_fd_by_id'?
../src/bpf/bpfd-manager.c:79:22: note: 'sym_bpf_map_get_fd_by_id' declared here
                fd = sym_bpf_map_get_fd_by_id(id);
                     ^
../src/bpf/bpfd-manager.c:134:16: error: implicit declaration of function 'sym_bpf_prog_get_next_id' is invalid in C99 [-Werror,-Wimplicit-function-declaration]
        while (sym_bpf_prog_get_next_id(id, &id) >= 0) {
               ^
../src/bpf/bpfd-manager.c:134:16: note: did you mean 'sym_bpf_map_get_next_id'?
../src/bpf/bpfd-manager.c:70:16: note: 'sym_bpf_map_get_next_id' declared here
        while (sym_bpf_map_get_next_id(id, &id) >= 0) {
               ^
../src/bpf/bpfd-manager.c:143:22: error: implicit declaration of function 'sym_bpf_prog_get_fd_by_id' is invalid in C99 [-Werror,-Wimplicit-function-declaration]
                fd = sym_bpf_prog_get_fd_by_id(id);
                     ^
../src/bpf/bpfd-manager.c:143:22: note: did you mean 'sym_bpf_prog_get_next_id'?
../src/bpf/bpfd-manager.c:134:16: note: 'sym_bpf_prog_get_next_id' declared here
        while (sym_bpf_prog_get_next_id(id, &id) >= 0) {
               ^
../src/bpf/bpfd-manager.c:150:21: error: implicit declaration of function 'sym_bpf_obj_get_info_by_fd' is invalid in C99 [-Werror,-Wimplicit-function-declaration]
                r = sym_bpf_obj_get_info_by_fd(fd, &prog_info, &prog_info_len);
                    ^
../src/bpf/bpfd-manager.c:163:65: error: use of undeclared identifier 'BPF_OBJ_NAME_LEN'
                        prog_map_names = new(char, nr_map_ids * BPF_OBJ_NAME_LEN);
                                                                ^
../src/bpf/bpfd-manager.c:168:37: error: variable has incomplete type 'struct bpf_prog_info'
                        prog_info = (struct bpf_prog_info) {
                                    ^~~~~~~~~~~~~~~~~~~~~~~~
../src/bpf/bpfd-manager.c:55:16: note: forward declaration of 'struct bpf_prog_info'
        struct bpf_prog_info prog_info = {};
               ^
../src/bpf/bpfd-manager.c:204:26: error: use of undeclared identifier 'BPF_TAG_SIZE'
                prog_tag[BPF_TAG_SIZE * 2] = '\0';
                         ^
../src/bpf/bpfd-manager.c:205:40: error: use of undeclared identifier 'BPF_TAG_SIZE'
                for (size_t j = 0; j < BPF_TAG_SIZE; ++j) {
                                       ^
../src/bpf/bpfd-manager.c:210:78: error: use of undeclared identifier 'BPF_OBJ_NAME_LEN'
                if (strlen(prog_info.name) == 0 || strlen(prog_info.name) >= BPF_OBJ_NAME_LEN - 1) {
                                                                             ^
../src/bpf/bpfd-manager.c:238:31: error: implicit declaration of function 'sym_btf__load_from_kernel_by_id' is invalid in C99 [-Werror,-Wimplicit-function-declaration]
                        btf = sym_btf__load_from_kernel_by_id(prog_info.btf_id);
                              ^
../src/bpf/bpfd-manager.c:239:29: error: implicit declaration of function 'sym_libbpf_get_error' is invalid in C99 [-Werror,-Wimplicit-function-declaration]
                        if (sym_libbpf_get_error(btf)) {
                            ^
../src/bpf/bpfd-manager.c:244:29: error: implicit declaration of function 'sym_btf__type_by_id' is invalid in C99 [-Werror,-Wimplicit-function-declaration]
                        t = sym_btf__type_by_id(btf, func_info.type_id);
                            ^
../src/bpf/bpfd-manager.c:250:44: error: implicit declaration of function 'sym_btf__name_by_offset' is invalid in C99 [-Werror,-Wimplicit-function-declaration]
                        prog_name = strdup(sym_btf__name_by_offset(btf, t->name_off));
                                           ^
fatal error: too many errors emitted, stopping now [-ferror-limit=]
20 errors generated.
```

poettering commented on src/bpf/bpfctl.c:
```diff
@@ -0,0 +1,220 @@
+                        r = table_add_many(map_table,
+                                   TABLE_STRING, p.name, TABLE_SET_ALIGN_PERCENT, 100,
+                                   TABLE_UINT64, p.type, TABLE_SET_ALIGN_PERCENT, 100,
+                                   TABLE_STRING, p.memlock, TABLE_SET_ALIGN_PERCENT, 100);
```
`r` must checked after this

poettering commented on src/bpf/bpfctl.c:
```diff
@@ -0,0 +1,220 @@
+        }
+        table_print(map_table, NULL);
+
+        return r;
```
please clean up error handling, i.e. we generally have the rule that functions should log about all their errors themselves, or that they should propagate them all (so that the caller can log). "main-function-like" functions should typically do the former, "library-like" the latter. The line between them is not always clear, but I'd say this is the former, i.e. it should log about all errors. (logging at debug level is always ok, and not considered logging in this case).

right now you handle one error with "log_oom()" (i.e. with proper logging), others you handle with debug logging, and even others you don't handle at all.

hence, please clean thus up, which probably means adding log_error_errno(…) or so everywhere

poettering commented on src/bpf/bpfd-manager.c:
```diff
@@ -0,0 +1,465 @@
+                map_name = strdup(map_info.name);
+                if (!map_name)
+                        return -ENOMEM;
+                hashmap_put(map_name_by_id, UINT32_TO_PTR(map_id), map_name);
```
error checking missing

poettering commented on src/bpf/bpfd-manager.c:
```diff
@@ -0,0 +1,465 @@
+                /* Store map names into hashmap for later lookup when iterating progs */
+                map_name = strdup(map_info.name);
+                if (!map_name)
+                        return -ENOMEM;
```
log_oom()

poettering commented on src/bpf/bpfd-manager.c:
```diff
@@ -0,0 +1,465 @@
+        if (r < 0)
+                return r;
+
+        memlock_line = find_line_startswith(fdinfo, "memlock:");
```
what does the field actually contain? a numeric value? if so, please parse it via safe_atou64() or similar. We generally prefer to pass around parsed data

poettering commented on src/bpf/bpfd-manager.c:
```diff
@@ -0,0 +1,465 @@
+                                continue;
+                        }
+                } else
+                        json_variant_append_array(&prog_and_map_list, v);
```
error handling

poettering commented on src/bpf/bpfd-manager.c:
```diff
@@ -0,0 +1,465 @@
+                                return -ENOMEM;
+                        prog_map_names = new(char, nr_map_ids * BPF_OBJ_NAME_LEN);
+                        if (!prog_map_names)
+                                return -ENOMEM;
```
log_oom()

poettering commented on src/bpf/bpfd-manager.c:
```diff
@@ -0,0 +1,465 @@
+                if (nr_map_ids > 0) {
+                        prog_map_ids = new(uint32_t, nr_map_ids);
+                        if (!prog_map_ids)
+                                return -ENOMEM;
```
log_oom()

poettering commented on src/bpf/bpfd-manager.c:
```diff
@@ -0,0 +1,465 @@
+                                prog_map_names[i] = ':';
+                                i++;
+                        }
+                        prog_map_names[i] = '\0';
```
just use this:

```c
if (strextend_with_separator(&prog_map_names, ":", map_name) < 0)
        return log_oom();
```

and thus use dynamic memory.

poettering commented on src/bpf/bpfd-manager.c:
```diff
@@ -0,0 +1,465 @@
+                /* Format prog tag */
+                prog_tag[BPF_TAG_SIZE * 2] = '\0';
+                for (size_t j = 0; j < BPF_TAG_SIZE; ++j)
+                        sprintf(&(prog_tag[2 * j]), "%02hhx", prog_info.tag[j]);
```
```c
prog_tag = hexmem(prog_info.tag, BPF_TAG_SIZE);
if (!prog_tag)
        return log_oom();
```

and this use dynamic memory.

in systemd we really care about keeping things easy to read. I doubt this is so performance sensitive that a malloc() here would actually hurt

poettering commented on src/bpf/bpfd-manager.c:
```diff
@@ -0,0 +1,465 @@
+                        sprintf(&(prog_tag[2 * j]), "%02hhx", prog_info.tag[j]);
+
+                /* Check if prog name is empty or might be truncated. If so, try to get full name from BTF */
+                if (strlen(prog_info.name) == 0 || strlen(prog_info.name) >= BPF_OBJ_NAME_LEN - 1) {
```
maybe use isempty(prog_info.name) for the first check

poettering commented on src/bpf/bpfd-manager.c:
```diff
@@ -0,0 +1,465 @@
+                        zero(prog_info);
+                        prog_info.nr_func_info = 1;
+                        prog_info.func_info_rec_size = func_info_rec_size;
+                        prog_info.func_info = PTR_TO_UINT64(&func_info);
```
please use compound initialization:

```c
prog_info = (struct bpf_prog_info) {
        .nr_func_info = 1,
        .func_info_rec_size = …
}
```

you get the idea.

poettering commented on src/bpf/bpfd-manager.c:
```diff
@@ -0,0 +1,465 @@
+        struct btf *btf = NULL;
+        char prog_tag[BPF_TAG_SIZE * 2 + 1];
+        _cleanup_hashmap_free_ Hashmap *map_name_by_id = NULL;
+        JsonVariant *prog_and_map_list = NULL;
```
i sense that some of these could be moved to an inner scope, couldn't they?

we generally want that variables are declared at the innermost scope whenever possible

(`int r;` is special however, that always gets outermost scope, it's our error variable of choice)

poettering commented on src/bpf/bpfd-manager.c:
```diff
@@ -0,0 +1,465 @@
+        if (r < 0)
+                log_warning_errno(r, "Failed to write '%s' to /proc/sys/kernel/bpf_stats_enabled, ignoring: %m", "0");
+
+
```
spurious double newline

poettering commented on src/bpf/bpfd-manager.c:
```diff
@@ -0,0 +1,465 @@
+
+        r = write_string_file("/proc/sys/kernel/bpf_stats_enabled", "1", WRITE_STRING_FILE_DISABLE_BUFFER);
+        if (r < 0)
+                log_warning_errno(r, "Failed to write '%s' to /proc/sys/kernel/bpf_stats_enabled, ignoring: %m", "1");
```
you you move this into some helper that just takes a bool and is called one with true and once with false?

poettering commented on src/bpf/bpfd.conf:
```diff
@@ -0,0 +1,16 @@
+# configuration file and all drop-ins located in /etc/.
+
+[Bpfd]
+#LogProgsAndMaps=yes
```
please don't abbreviate config settings needlessly it makes them so cryptic... i.e. Programs rather than Profs please

arthurshau commented:
Addressed comments - most notably, error handling throughout the code should be in a much nicer state now. Not sure about why the CI is failing though...would appreciate any feedback on how best to fix that.

arthurshau commented:
Rebase.

Also here's a picture of what ```bpfctl list``` looks like now with the table formatting.

![image](https://github.com/systemd/systemd/assets/13985155/9a4c4434-e4c5-4267-8644-acb8ea02bc23)

poettering commented on src/bpf/bpfctl.c:
```diff
@@ -0,0 +1,238 @@
+/* SPDX-License-Identifier: LGPL-2.1-or-later */
```
can you add bpfctl in a 2nd commit please?

poettering commented on src/bpf/bpfctl.c:
```diff
@@ -0,0 +1,238 @@
+        JsonVariant *i = NULL, *o = NULL, *m = NULL, *v = NULL;
+        int r;
+
+        map_table = table_new("map_name", "map_type", "memlock");
```
hmm, these are supposed to be short but human-directed column names, i.e. should probably be "name" and "type" just.

poettering replied:
the "memlock" column, what's that? locked memory?

shouldn't this just be called "memory"?

arthurshau replied:
Hmm, the thing is, the line where this info comes from, in /proc/self/fdinfo/<fd>, itself labels the value as "memlock." So we might want to maintain consistency?

There is also the issue that this memlock value sometimes varies from the real memory usage (such as for non-preallocated bpf maps, which may increase or decrease real memory size dynamically), so labelling it as just 'memory' may be misleading.

If just 'memlock' is too vague though, I am open to renaming the column to 'locked memory' or something similar.

poettering replied:
so we generally want to expose a more "human" interface to people that procfs would do it. that said, if memlock is really the better term, then i am fine with it.

poettering commented on src/bpf/bpfctl.c:
```diff
@@ -0,0 +1,238 @@
+                        r = table_add_many(map_table,
+                                   TABLE_STRING, p.name, TABLE_SET_ALIGN_PERCENT, 100,
+                                   TABLE_UINT64, p.type, TABLE_SET_ALIGN_PERCENT, 100,
+                                   TABLE_UINT64, p.memlock, TABLE_SET_ALIGN_PERCENT, 100);
```
hmm, if "memlock" refers to bytes of memory, then should use TABLE_SIZE as field type

poettering commented on src/bpf/bpfctl.c:
```diff
@@ -0,0 +1,238 @@
+                                { "tag", JSON_VARIANT_STRING, json_dispatch_string, offsetof(struct BpfProgramData, tag), JSON_MANDATORY},
+                                { "map_names", JSON_VARIANT_ARRAY, json_dispatch_strv, offsetof(struct BpfProgramData, map_names), 0},
+                                { "run_time_ns", JSON_VARIANT_UNSIGNED, json_dispatch_uint64, offsetof(struct BpfProgramData, run_time_ns), JSON_MANDATORY},
+                                { "run_count", JSON_VARIANT_UNSIGNED, json_dispatch_uint64, offsetof(struct BpfProgramData, run_cnt), JSON_MANDATORY},
```
lease align table (here and everywhere)

poettering commented on src/bpf/bpfctl.c:
```diff
@@ -0,0 +1,238 @@
+        if (!map_table)
+                return log_oom();
+
+        prog_table = table_new("program_name", "program_type", "memlock", "program_tag", "program_map_names", "run_time_ns", "run_cnt");
```
as above, drop the redundant "program_" prefix, and use spaces rather than under scores.

goal is short human friendly names, that are not too cryptic

poettering commented on src/bpf/bpfctl.c:
```diff
@@ -0,0 +1,238 @@
+                if (!v)
+                        return log_error_errno(SYNTHETIC_ERRNO(EINVAL), "Failed to find data field of json variant");
+
+                if (streq(json_variant_string(m), "map")) {
```
maybe use streq_ptr() at least, just in case the field is not actually a string

poettering commented on src/bpf/bpfctl.c:
```diff
@@ -0,0 +1,238 @@
+
+                        r = table_add_many(map_table,
+                                   TABLE_STRING, p.name, TABLE_SET_ALIGN_PERCENT, 100,
+                                   TABLE_UINT64, p.type, TABLE_SET_ALIGN_PERCENT, 100,
```
for display this really should be translated to a string. use the stuff from string-table.h for that (unless libbpf has a helper call for translating type numbers to strings, in which case better use that)

arthurshau replied:
There are libbpf helpers for it, called libbpf_bpf_map_type_str(...), and libbpf_bpf_prog_type_str(...).

I just added those to bpf-dlopen, but for some reason, when calling them from bpfctl, the program coredumps. In fact, all the dlsym'd libbpf functions coredump when called from bpfctl. I'm not exactly sure why - they work perfectly fine in bpfd-manager.c. I've commented out the relevant code in bpfctl.c that causes the coredump.

arthurshau replied:
Resolved

poettering commented on src/bpf/bpfctl.c:
```diff
@@ -0,0 +1,238 @@
+                        static const JsonDispatch dispatch_table[] = {
+                                { "name",  JSON_VARIANT_STRING, json_dispatch_string, offsetof(struct BpfProgramData, name),  JSON_MANDATORY },
+                                { "type", JSON_VARIANT_UNSIGNED, json_dispatch_uint64, offsetof(struct BpfProgramData, type), JSON_MANDATORY},
+                                { "memlock", JSON_VARIANT_UNSIGNED, json_dispatch_uint64, offsetof(struct BpfProgramData, memlock), JSON_MANDATORY},
```
here and everywhere: when you use json_dispatch_uint64() then please set the 2nd column here to: _JSON_VARIANT_TYPE_INVALID. That has the effect that the field type check is dropped. This is a good idea, because JSON famously is not 64bit clean, because of its JS heritage. i.e. integers are only safe in the 53bit range. This means, when inserting integers in the 53…64 bit range various implementations format them as string field. 

json_dispatch_uint64() will handle this transparently for you, if you let it. But that only works, if you allow both integers and strings through. But via fixating the type to JSON_VARIANT_UNSIGNED here you do not allow that.

hence, whenever you use json_dispatch_uint64() (and similar number parsers) make sure to set the 2nd column to _JSON_VARIANT_TYPE_INVALID. json_dispatch_uint64() internally validates the types anyway, hence you lose nothing

poettering commented on src/bpf/bpfd.conf:
```diff
@@ -0,0 +1,16 @@
+# recommended. Defaults can be restored by simply deleting the main
+# configuration file and all drop-ins located in /etc/.
+
+[Bpfd]
```
i'd really call this section [BPF]. 

We never add the d suffix in the section names of our current config files, hence I doubt we should do so here. And the lowercasing of BPF is a bit weird I think

poettering commented on src/bpf/bpfd.c:
```diff
@@ -0,0 +1,42 @@
+#include "daemon-util.h"
+#include "bpfd-manager.h"
+#include "main-func.h"
+#include "log.h"
```
please sort alphabetically (here and everywhere)

poettering commented on src/bpf/bpfd-manager.h:
```diff
@@ -0,0 +1,26 @@
+#include "varlink.h"
+
+/* Polling interval for BPF timer  */
+#define BPF_TIMER_INTERVAL_USEC 60 * USEC_PER_SEC /* 1 minute */
```
please enclose in (), so that it's safe to use in arbitrary mathematical expressions

poettering commented on src/bpf/bpfctl.c:
```diff
@@ -0,0 +1,238 @@
+                return log_oom();
+
+        printf("%s [OPTIONS...] COMMAND\n\n"
+               "%sQuery running bpf programs and maps.%s\n"
```
i'd put BPF in upper case here

poettering commented on src/bpf/bpfctl.c:
```diff
@@ -0,0 +1,238 @@
+        if (r < 0)
+                return log_error_errno(r, "Failed to output map table: %m");
+
+        r = table_print(prog_table, NULL);
```
please use table_print_with_pager()

poettering commented on src/bpf/bpfctl.c:
```diff
@@ -0,0 +1,238 @@
+        uint64_t memlock;
+};
+
+static void bpf_prog_data_done(BpfProgramData *p) {
```
please name the functions just like the types they destroy, i.e. bpf_program_data_done() here. those three characters won't hurt

poettering commented on src/bpf/bpfd-manager.c:
```diff
@@ -0,0 +1,489 @@
+#include "hexdecoct.h"
+#include "parse-util.h"
+#include "sd-daemon.h"
+#include "sd-messages.h"
```
the sd-*.h headers should be included first (as they are the apis we define that are accessible to other projects too, and thus are more "public" then the other headers we have. and place an empty line between them and the other include lines

poettering commented on src/bpf/bpfd-manager.c:
```diff
@@ -0,0 +1,489 @@
+
+        const char *value = enable ? "1" : "0";
+
+        r = write_string_file("/proc/sys/kernel/bpf_stats_enabled", value, WRITE_STRING_FILE_DISABLE_BUFFER);
```
change 2nd param to `one_zero(enable)`

poettering commented on src/bpf/bpfd-manager.c:
```diff
@@ -0,0 +1,489 @@
+        JsonVariant *prog_and_map_list = NULL;
+        int r;
+
+        map_name_by_id = hashmap_new(NULL);
```
we nowadays don't allocate hashmaps explicitly anymore, but prefer it if they are allocated the moment you add the first entry to it. use hashmap_ensure_put() as a call that allocates a hashmap if not done yet, and then inserts an entry to it.

poettering commented on src/bpf/bpfd-manager.c:
```diff
@@ -0,0 +1,489 @@
+                char *map_name = NULL;
+                JsonVariant *v;
+
+                zero(map_info);
```
you already zeroed it above via `= {}`

poettering commented on src/bpf/bpfd-manager.c:
```diff
@@ -0,0 +1,489 @@
+
+        char *map_name;
+        HASHMAP_FOREACH(map_name, map_name_by_id)
+                free(map_name);
```
please use the hash_ops stuff so that the when the hashmap is freed it automatically frees its remaining entries.

arthurshau replied:
I used _cleanup_hashmap_free_free_ instead of hash_ops, since I couldn't find one that frees the values only. This hashmap simply maps from a uint32_t to a string, so I think that should be sufficient.

arthurshau replied:
Also, if you wouldn't mind double checking that I'm freeing the Hashmap and JsonVariant stuff properly, I would really appreciate it. I'm not confident whether or not I'm doing it correctly, especially with passing JsonVariant pointers around.

poettering commented on src/bpf/bpfd-manager.c:
```diff
@@ -0,0 +1,489 @@
+                _cleanup_close_ int fd = -EBADF;
+                uint64_t memlock = 0;
+                char *map_name = NULL;
+                JsonVariant *v;
```
the reference is leaked. You need to make this `_cleanup_(json_varian_unrefp) JsonVariant *v = NULL;`

poettering commented on src/bpf/bpfd-manager.c:
```diff
@@ -0,0 +1,489 @@
+                JsonVariant *v = NULL;
+                _cleanup_strv_free_ char **prog_map_names = NULL;
+
+                zero(prog_info);
```
as above

poettering commented on src/bpf/bpfd-manager.c:
```diff
@@ -0,0 +1,489 @@
+                        prog_name = strdup(sym_btf__name_by_offset(btf, t->name_off));
+                        if (!prog_name) {
+                                return log_oom();
+                        }
```
as per coding style, please no {} around single line if blocks.

poettering commented on src/bpf/bpfd-manager.c:
```diff
@@ -0,0 +1,489 @@
+        return 0;
+}
+
+static int fetch_bpf_maps_and_progs(Varlink *link) {
```
any reason why this is a single varlink method that gets both? feels a bit strange? why not make it two calls?

arthurshau commented on src/bpf/bpfd-manager.c:
```diff
@@ -0,0 +1,489 @@
+        (void) mkdir_p("/run/systemd/bpf", 0755);
+        socket_path = "/run/systemd/bpf/io.systemd.Bpf";
+
+        r = varlink_server_listen_address(m->varlink_server, socket_path, 0666);
```
I changed it to 0600 instead, should be sufficient?

arthurshau commented:
Addressed comments. Notable changes include:
- Addition of bpfctl is now split into a second commit.
- Varlink API has been split into two methods, Bpf.DescribeMaps and Bpf.DescribePrograms. bpfctl was updated to reflect this - you can call ```bpfctl list-maps``` to list maps only or ```bpfctl list-progs``` to list programs only. ```bpfctl list-all``` (default) now prints all.
- Similar to above, fetching of maps and progs was split into 2 functions internally in bpfd-manager.c. They also now get logged separately to the journal.

arthurshau commented:
rebased

keszybz commented on src/bpf/bpfd-manager.c:
```diff
@@ -0,0 +1,533 @@
+static void toggle_bpf_stats_enabled(bool enable) {
+        int r;
+
+        const char *value = enable ? "1" : "0";
```
This definition can be dropped.

keszybz commented on src/bpf/bpfd-manager.c:
```diff
@@ -0,0 +1,533 @@
+
+        r = write_string_file("/proc/sys/kernel/bpf_stats_enabled", one_zero(enable), WRITE_STRING_FILE_DISABLE_BUFFER);
+        if (r < 0)
+                log_error_errno(r, "Failed to write '%c' to /proc/sys/kernel/bpf_stats_enabled, ignoring: %m", *value);
```
Just use `%s` and `one_zero(enable)` again here.

keszybz commented on src/bpf/bpfd-manager.c:
```diff
@@ -0,0 +1,533 @@
+
+        r = safe_atou64(memlock_line, &v);
+        if (r < 0)
+                return log_error_errno(r, "Failed to convert memlock string to unsigned integer, ignoring: %m");
```
```C
"Failed to parse memlock value '%s', ignoring: %m", memlock_line
```

keszybz commented on src/bpf/bpfd-manager.c:
```diff
@@ -0,0 +1,533 @@
+        if (r < 0)
+                return log_error_errno(r, "Failed to convert memlock string to unsigned integer, ignoring: %m");
+
+        *ret_memlock = v;
```
`v` is not needed. `safe_atou64` only writes the output on success, so just put `ret_memlock` directly in the call.

keszybz commented on src/bpf/bpfd-manager.c:
```diff
@@ -0,0 +1,533 @@
+        int r;
+
+        /* Iterate through all attached BPF progs*/
+        id = 0;
```
This was already initialized to 0 above.

keszybz commented on src/bpf/bpfd-manager.c:
```diff
@@ -0,0 +1,533 @@
+                /* Grab the FDs using their IDs*/
+                fd = sym_bpf_prog_get_fd_by_id(id);
+                if (fd < 0) {
+                        log_error_errno(errno, "Can't get fd for prog id %u, ignoring: %m", id);
+                        continue;
+                }
+
+                /* Fetch bpf_prog_info using fd */
+                r = sym_bpf_obj_get_info_by_fd(fd, &prog_info, &prog_info_len);
+                if (r < 0) {
+                        log_error_errno(r, "Can't get bpf prog info for fd %i, ignoring: %m", fd);
```
Bleh, is the error returned using `errno` in one case and the return value in the other?

arthurshau replied:
Ah my mistake, it should be using errno in both cases. Both functions return zero on success, and on error, -1 is returned and errno is set appropriately. Fixed everywhere.

keszybz commented on src/bpf/bpfd-manager.c:
```diff
@@ -0,0 +1,533 @@
+                        if (!prog_map_ids)
+                                return log_oom();
+
+                        zero(prog_info);
+                        prog_info = (struct bpf_prog_info) {
```
It doesn't make sense to zero the struct, if it's about to be immediately overridden anyway.

keszybz commented on src/bpf/bpfd-manager.c:
```diff
@@ -0,0 +1,533 @@
+
+        r = varlink_server_listen_address(m->varlink_server, socket_path, 0600);
+        if (r < 0)
+                return log_error_errno(r, "Failed to bind to varlink socket: %m");
```
Please include the path to the socket in the error message, that is the most important piece of information for a user.

keszybz commented on src/bpf/bpfd-manager.c:
```diff
@@ -0,0 +1,533 @@
+        sd_event_unref(m->event);
+        varlink_server_unref(m->varlink_server);
+
+        toggle_bpf_stats_enabled(/*enable:*/false);
```
For consistency with the rest of the code, please use `/* enable= */ false`.

keszybz commented on src/bpf/bpfd-manager.c:
```diff
@@ -0,0 +1,533 @@
+        if (r < 0)
+                return log_error_errno(r, "Couldn't dlopen_bpf: %m");
+
+        toggle_bpf_stats_enabled(/*enable:*/true);
```
Same here.

keszybz commented on src/bpf/bpfd.c:
```diff
@@ -0,0 +1,42 @@
+
+        r = manager_start(m);
+        if (r < 0)
+                return log_error_errno(r, "Failed to start up daemon: %m");
```
"Failed to start daemon" sounds better.

keszybz commented on src/bpf/bpfctl.c:
```diff
@@ -0,0 +1,298 @@
+        if (!prog_table)
+                return log_oom();
+
+        r = varlink_connect_address(&vl, "/run/systemd/bpf/io.systemd.Bpf");
```
This string is already defined in the daemon. I would prefer to have a `#define BPFD_VALINK_ADDRESS "…"` in a header somewhere.

arthurshau replied:
Sure, I defined it in bpfd-manager.h.

keszybz commented on src/bpf/bpfctl.c:
```diff
@@ -0,0 +1,298 @@
+
+        r = varlink_connect_address(&vl, "/run/systemd/bpf/io.systemd.Bpf");
+        if (r < 0)
+                return log_error_errno(r, "Unable to connect to /run/systemd/bpf/io.systemd.Bpf: %m");
```
%s here. It's possibly a bit less efficient, but also less prone to errors.

keszybz commented on src/bpf/bpfctl.c:
```diff
@@ -0,0 +1,298 @@
+        assert(argc >= 0);
+        assert(argv);
+
+        while ((c = getopt_long(argc, argv, "hasln:", options, NULL)) >= 0) {
```
`{}` can be dropped here.

keszybz reviewed (commented):
This looks pretty nice. Some stylistic comments.

arthurshau commented:
Addressed @keszybz's comments and rebased.

arthurshau commented:
![image](https://github.com/systemd/systemd/assets/13985155/ea9991ee-3c9e-4f1e-b604-7da28713b19c)

arthurshau commented:
I've pushed a commit that renames ```bpfd``` to ```bpfstatd```. This emphasizes its use-case as a daemon solely meant for collecting and logging information (stats) about BPF objects loaded in the kernel. Likewise, ```bpfctl``` has changed to ```bpfstatctl```. Please let me know if I've missed anything.

jordalgo commented:
Awesome! Looking forward to this getting merged.

anitazha commented:
After some internal discussions we've decided to close this PR until we have a more fleshed out design goals for BPFd. The metrics use case alone is not compelling enough and it would be better for the community if we submit something that is more complete that people want (e.g. revisiting BPF program management and delegation).
